### DISCUSSION TITLE
Is there a way to do SPA/master-detail routing for a single page?

### DISCUSSION THREAD

**Discussion by Svish** [vercel/next.js#22217] [Help]
Say you have an admin page with a list of products on the left, and when selecting one of them, you want product details to show up on the right. And additionally, you want each product to be linkable. So navigating to `/admin/products` would show the list of products, but no product selected yet, and `/admin/products/123` would show the same list and product `123` already selected and details loaded.

I could of course try to pull a `<ProductList>` component out, fetch its data in _both_ `pages/admin/products.js` _and_ `pages/admin/products/[pid].js`, but that "feels wrong", and I assume it would also cause a reload of the page, or at least a re-mount of the `<ProductList>` component when switching products? If the `<ProductList>` included a search and a filter for example, that shouldn't be "dropped" when selecting a product.

Maybe I'm just overthinking this, but wondering how one would write a page like this in Next.js without causing a mess. It would become even more of a mess if you for example, under each product detail had some tabs with their own data, which should also be linkable. For example say `/admin/products/123/history` would show something like this:

![image](https://user-images.githubusercontent.com/142162/108056933-3c865d80-7052-11eb-872a-d7b31f20ec4c.png)

Is there something built-in to Next.js to do this? Or would you need to use something like an [optional catch-all route](https://nextjs.org/docs/routing/dynamic-routes#optional-catch-all-routes), like `/admin/products/[[...pid]].js`, and then pull in `react-router` to use it on this specific page or something? Would that even work?

Maybe I'm just overthinking this greatly, but yeah... I need some help understanding what a setup like this could look like in Next.js. With `create-react-app` and `react-router` it would just be a matter of "going nuts" with nested routes, but since `Next.js` uses file-based routing, I'm not sure how to do a similar setup here.

**Comment by Svish** [comment]
Found [this blog post](https://colinhacks.com/essays/building-a-spa-with-nextjs), and seems "the way" is indeed to pull inn e.g. `react-router`, and do some hoop jumping via Next.js redirects and such. 🤔

I suppose that could work, but makes me wonder: Wouldn't it be much better if Next.js had a built-in feature to allow easily mixing in setups like these? Any plans for that?

**Reply by kiecooboi** [reply]
Hopping so.

**Comment by johnculviner** [comment]
I think the solution here is the use layout(s): 

https://nextjs.org/docs/basic-features/layouts

and just normal routing as you've indicated.

**Comment by sergio-milu** [comment]
I was wondering how could we achieve this if we have a responsive design, in mobile the first the list and in a second page the detail, and in desktop both in the same page (as in the first post), is this possible?

**Reply by Svish** [reply]
Probably possible with the new app directory.

**Reply by sergio-milu** [reply]
any idea how? I've been playing around and I could achieve it

**Reply by Svish** [reply]
With app directory you can create nested pages. So I would create the details page a nested child page of the list page. And then you could for example just minimise or hide the list on the details page.
